File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.py
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        edgeDict = collections.defaultdict(list)
        for vertex, edge, weight in flights:
            edgeDict[vertex].append((edge, weight))

        costs = [float('inf') for _ in range(n)]
        costs[src] = 0
        
        for _ in range(k + 1):
            temp = costs.copy()
            for vertex in range(n):
                for edge, weight in edgeDict[vertex]:
                    temp[edge] = min(temp[edge], costs[vertex] + weight)
            costs = temp
        
        if temp[dst] != float('inf'):
            return temp[dst]
        else:
            return -1

File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int K) {
        //bellman ford calculates the min path for upto i edges in path at the ith iteration.
        vector<int> dist(n,1e9);
        dist[src] = 0;
        for(int k=0;k<K+1;k++){
            vector<int> temp(dist);
            for(auto& v:flights){
                if( dist[v[0]]!=1e9 && temp[v[1]]>dist[v[0]]+v[2]){
                    temp[v[1]] = dist[v[0]]+v[2];
                }
            }
            dist=temp;
        }
        if(dist[dst]==1e9){return -1;}
        return dist[dst];
    }
};